lighthouse-core/computed/critical-request-chains.js

```diff
@@ -66,6 +66,68 @@ class CriticalRequestChains {
     return ['VeryHigh', 'High', 'Medium'].includes(request.priority);
   }
 
+  /**
+   * Create a tree of critical requests.
+   * @param {LH.Artifacts.NetworkRequest} mainResource
+   * @param {LH.Gatherer.Simulation.GraphNode} graph
+   * @return {LH.Artifacts.CriticalRequestNode}
+   */
+  static extractChainsFromGraph(mainResource, graph) {
+    /** @type {LH.Artifacts.CriticalRequestNode} */
+    const rootNode = {};
+
+    /**
+     * @param {LH.Artifacts.NetworkRequest[]} path
+     */
+    function addChain(path) {
+      let currentNode = rootNode;
+
+      for (const record of path) {
+        if (!currentNode[record.requestId]) {
+          currentNode[record.requestId] = {
+            request: record,
+            children: {},
+          };
+        }
+
+        currentNode = currentNode[record.requestId].children;
+      }
+    }
+
+    // By default `traverse` will discover nodes in BFS-order regardless of dependencies, but
+    // here we need traversal in a topological sort order. We'll visit a node only when its
+    // dependencies have been met.
+    const seenNodes = new Set();
```

Comment:
Maybe we should pull this into graph, and make a `topologicalTraverse` ?

Reply:
Are you thinking there's another use for this? If we plan on using it again anywhere, sounds good :)

Reply:
I don't know yet, but it does seem like a basic functionality of a graph.

Reply:
I guess my hesitation is that this is the first time we've needed topological sort traversal in ~3 years of working with our graphs, so I'm skeptical it will come up again soon enough to justify the preemptive DRY work (and send this logic further away if we need to modify it in the next immediate step of the CRC refactor).